controllers/medusa/medusatask_controller.go

```diff
@@ -0,0 +1,470 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package medusa
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/go-logr/logr"
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	medusav1alpha1 "github.com/k8ssandra/k8ssandra-operator/apis/medusa/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/config"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/medusa"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+)
+
+// MedusaTaskReconciler reconciles a MedusaTask object
+type MedusaTaskReconciler struct {
+	*config.ReconcilerConfig
+	client.Client
+	Scheme *runtime.Scheme
+	medusa.ClientFactory
+}
+
+// +kubebuilder:rbac:groups=medusa.k8ssandra.io,namespace="k8ssandra",resources=medusatasks,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=medusa.k8ssandra.io,namespace="k8ssandra",resources=medusatasks/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=medusa.k8ssandra.io,namespace="k8ssandra",resources=medusatasks/finalizers,verbs=update
+// +kubebuilder:rbac:groups=medusa.k8ssandra.io,namespace="k8ssandra",resources=cassandradatacenters,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",namespace="k8ssandra",resources=pods;services,verbs=get;list;watch
+
+func (r *MedusaTaskReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("MedusaTask", req.NamespacedName)
+
+	logger.Info("Starting reconciliation for MedusaTask")
+
+	// Fetch the MedusaTask instance
+	instance := &medusav1alpha1.MedusaTask{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		logger.Error(err, "Failed to get MedusaTask")
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+	}
+
+	task := instance.DeepCopy()
+
+	// First check to see if the task is already in progress
+	if !task.Status.StartTime.IsZero() {
+		// If there is anything in progress, simply requeue the request
+		if len(task.Status.InProgress) > 0 {
+			logger.Info("Tasks already in progress")
+			return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+		}
+
+		logger.Info("task complete")
+
+		// Set the finish time
+		// Note that the time here is not accurate, but that is ok. For now we are just
+		// using it as a completion marker.
+		patch := client.MergeFrom(task.DeepCopy())
```

Comment:
task is already DeepCopy of instance. Do we really need double copy?

Reply:
I'm following the conventions we've used throughout the project when patching CRs.
Do you think we're wrong to do so?
I seem to remember that patches wouldn't be applied properly without this, but I could be wrong.